### ISSUE TITLE
RandomShuffle does not work properly while using ops.TFRecordReader

### ISSUE THREAD

**Issue by KleinXin** [NVIDIA/DALI#2688]
I am now doing  a classification job. So I am trying to read images and labels from tfrecords. The reading of tfrecords works fine because I tried to output the pytorch tensors as image during trainning. However, I found that RandomShuffle does not work properly while using ops.TFRecordReader.

First I used ops.FileReader to read images and labels from a folder and set random_shuffle as True. Then I used  ops.TFRecordReader and also set random_shuffle as True.

The codes I used are as below

```
import os

from nvidia.dali.pipeline import Pipeline
import nvidia.dali.ops as ops
import nvidia.dali.types as types
import nvidia.dali.tfrecord as tfrec
from nvidia.dali.plugin.pytorch import DALIGenericIterator

import cv2
import numpy as np

class CommonPipeline(Pipeline):
    def __init__(self, batch_size, num_threads, device_id,crop_size=448,img_size=512,randomFlip=True):
        super(CommonPipeline, self).__init__(batch_size, 
                                             num_threads, 
                                             device_id)
                                             
        self.randomFlip = randomFlip
        self.crop_size = crop_size
        self.img_size = img_size
        self.decode = ops.ImageDecoder(device = "mixed", output_type = types.RGB)
        self.resize = ops.Resize(device = "gpu",
                                 interp_type = types.INTERP_LINEAR)
        self.cmn = ops.CropMirrorNormalize(device = "gpu",
                                            dtype = types.FLOAT,
                                            crop = (crop_size,crop_size),
                                            mean = [123, 117, 104],
                                            std = [58.395, 57.12, 57.375])
        #self.uniform = ops.Uniform(range = (0.0, 1.0))
        #self.resize_rng = ops.Uniform(range = resizeRange)
        self.coin = ops.CoinFlip(probability=0.5)

    def base_define_graph(self, inputs, labels):
        images = self.decode(inputs)
        images = self.resize(images, 
                             resize_x = self.img_size,
                             resize_y = self.img_size)
        if self.randomFlip == True:
           output = self.cmn(images,
                             mirror = self.coin())
        else:
           output = self.cmn(images)
                          
        return (output, labels)

class FolderReadPipeline(CommonPipeline):
        def __init__(self, image_dir,
                           batch_size, 
                           num_threads, 
                           device_id,
                           crop_size=448,
                           image_size=512,
                           randomFlip=True,
                           random_shuffle=True):
            super(FolderReadPipeline, self).__init__(batch_size, num_threads, device_id,crop_size,image_size,randomFlip)
            self.input = ops.FileReader(file_root = image_dir,random_shuffle=random_shuffle)

        def define_graph(self):
            images, labels = self.input(name="Reader")
            return self.base_define_graph(images, labels)

class TFRecordPipeline(CommonPipeline):
    def __init__(self, tfrecord_path,
                       tfrecord_idx_path,
                       batch_size,
                       num_threads,
                       device_id,
                       crop_size=448,
                       image_size=512,
                       randomFlip=True,
                       random_shuffle=True):
        super(TFRecordPipeline, self).__init__(batch_size, num_threads, device_id,crop_size,image_size,randomFlip)
        self.input = ops.TFRecordReader(path = tfrecord_path,
                                        index_path = tfrecord_idx_path,
                                        features = {"image" : tfrec.FixedLenFeature((), tfrec.string, ""),
                                                    "label": tfrec.FixedLenFeature([1], tfrec.int64,  -1)
                                        },
                                        random_shuffle=random_shuffle)

    def define_graph(self):
        inputs = self.input(name="Reader")
        images = inputs["image"]
        labels = inputs["label"]
        return self.base_define_graph(images, labels) 
```

The classification job works fine when I used ops.FileReader. However, when ops.TFRecordReader is used, the accuracy remains about 20% for about 8 epochs and does not seems to change. So I stopped trainning process.

After that I tried to save images and lables during trainning, both of them are corect. But I found random shuffle does not work properly when ops.TFRecordReader is used. The lables of each batch are saved in one line in a txt file. It can be seen from the image below that dataset seems not be shuffled. The batch size is 36.

![image](https://user-images.githubusercontent.com/13084622/108254904-7d878a80-7196-11eb-86b2-2c05da12f4cd.png)

Compared to using ops.TFRecordReader, I also saved images and labels during trainning when ops.FileReader is used. The result is shown below. The random shuffle works fine in this case.

![image](https://user-images.githubusercontent.com/13084622/108255168-dc4d0400-7196-11eb-81aa-c2e068bd44d7.png)

I think the reason the trainnning of classification job can not work fine is because of the random shuffle. So I will try to shuffle the images while generating the tfrecords.

Any help is appreciated!

**Comment by JanuszL** [comment]
Hi,
DALI uses an internal buffer of the default 1024 size (you change it using `initial_fill`)  to generate random batches. Samples are read one by one from the dataset to the buffer and then the buffer is randomly sampled to form a batch. In the case of file reader, the data set is preshuffled once at the beginning of the training so you won't get label values from the beginning of the dataset. In the case of the TFRecord it is expected (as it is a sequentially accessed data container) that it is preshuffled already, like in [this example](https://github.com/visipedia/tfrecords/blob/master/create_tfrecords.py#L365).

**Comment by KleinXin** [comment]
> Hi,
> DALI uses an internal buffer of the default 1024 size (you change it using `initial_fill`) to generate random batches. Samples are read one by one from the dataset to the buffer and then the buffer is randomly sampled to form a batch. In the case of file reader, the data set is preshuffled once at the beginning of the training so you won't get label values from the beginning of the dataset. In the case of the TFRecord it is expected (as it is a sequentially accessed data container) that it is preshuffled already, like in [this example](https://github.com/visipedia/tfrecords/blob/master/create_tfrecords.py#L365).

As you suggestted, I do the random shuffle during the generating of tfrecords. The trainning process seems work fine. 

In fact, I generating the reading pipline every epoch. I think maybe it is better to use ops.FileReader because images are shuffled every epoch. However, when ops.TFRecordReader is used, the sequence of images remains nearly the same every epoch.

Let's wait for the results of trainning using ops.TFRecordReader.

**Comment by JanuszL** [comment]
>  ops.FileReader because images are shuffled every epoch.

The global shuffling happens only once during the pipeline construction, after that if uses the same method to do randomness as in TFRecordReader. If you want to shuffle each epoch use `shuffle_after_epoch`.

**Comment by KleinXin** [comment]
> > ops.FileReader because images are shuffled every epoch.
> 
> The global shuffling happens only once during the pipeline construction, after that if uses the same method to do randomness as in TFRecordReader. If you want to shuffle each epoch use `shuffle_after_epoch`.

The results of using ops.TFRecordReader are nearly the same as ops.FileReader, so I think it does not have some very large problem until now.

The process of my trainning codes are as followed

```
for epoch in range(num_epochs):

      construct trainning tfrecord reading pipline together with random shuffle(set random_shuffle param of ops.TFRecordReader to be true)

      do the trainning

      construct validation tfrecord reading pipline

      do the validation
```

In this process, how can I do the shuffle to the tfrecord every epoch instead of doing random shuffle during the generation of tfrecord?

'**_shuffle_after_epoch_**' as you said, is it a param of  ops.TFRecordReader?

**Comment by KleinXin** [comment]
> @KleinXin ,
> You should not rebuild your pipeline each epoch. Doing that doesn't improve the randomness.
> 'shuffle_after_epoch' works only in FileReader, COCOReader, and NumpyReader. As I said, readers for containers (Caffe LMDB, TFRecord, and RecordIO) rely on the speed of sequential access.
> I recommend checking our [TensorFlow ResNet50 example](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/use_cases/tensorflow/resnet-n/README.html). It converges with the expected accuracy.
> What you can do is:
> disable shuffling
> output labels from the reader
> if for the first few batches you get the same label you mean that your data set is not well suitable to use with DALI. The best you can do in your case is to increase the `initial_fill` argument value by a factor of 10 or even more from the default value

thx! I will try to increase the initial_fill param to 10240 without randome shuffle during the generation of tfrecords.

**Comment by KleinXin** [comment]
I did two tests. The results are shown below, offline means random shuffle is performed during tfrecords generation, online means random shuffle is performed during trainning by setting initial_fill param to 10240. It shows that online+offline random shuffle is better on my dataset.

Thank you for your help!

**Comment by JanuszL** [comment]
@KleinXin I'm glad you finally made it work. As I said DALI expects that TFRecord is shuffled offline already to make sure that the internal shuffling buffer DALI has it not filled with samples only from 1-2 classes at the beginning.
